### ISSUE TITLE
Migration to Github Actions

### ISSUE THREAD

**Issue by fsimonis** [precice/precice#713]
GitHub Actions gives us the possibility to run the tests in docker containers. This now opens the question on what to test. Below a list of things we should consider when building our test Matrix.

Questions to the reader:
* What else has to be tested?
* What can be removed from the test matrix?

## preCICE Feature Configurations
* PythonActions - This is a pure add-on, so let's always enable it.
* Fortran/C Bindings - Also a pure add-on, so let's always enable it.
* MPICommunication - This is required to run the actual test suite after #710 
* PETScMapping - This is not a pure add-on as both paths run code. This will be even more significant once we get a fallback implementation for parallel RBF.

Configurations to test (nothing changed):

 MPI | PETSc
 --- | ---
Off | Off
On | Off
On | On

## Build Type
* Debug - Enables assertions and debug output.
* Release - Optimized build which we did not test so far. (we should)

## Compiler
* GCC - This is what the vast majority of our users use.
* Clang - This is what some of the developers use. So we probably don't need to explicitly test this anymore.
* Intel - Significant for clusters, but cumbersome to integrate into testing. Let's ignore this for now.

## MPI Implementations
* Open MPI - default MPI on ubuntu / Arch
* MPICH - no precompiled packages available. We would have to build the dependencies from source. Spack could help.
* Intel MPI - cumbersome to setup. Let's ignore this.

## Systems
* Ubuntu 16.04 LTS - supported until 2021, possibly not worth it.
* Ubuntu 18.04 LTS - supported until 2023, our current baseline for dependencies
* Ubuntu 20.04 LTS - supported until 2025
* Arch Linux - rolling release, thus everything is always at the latest release. Official petsc package missing.
* CentOS 7 - frequently pops up with users.

## Dependency versions:
* Latest release (covered by Arch)
* Oldest supported release (covered by Ubuntu 18.04 LTS)

# Build Matrix

What | Options
--- | ---
Features | 3
Build Type | 2
Compiler | 1
MPI Implementations | 2
Systems | 4
**Product** | **48**

**Comment by uekerman** [comment]
Looks good in my opinion. Thanks for the overview. 
I guess we can still decrease the number a bit:
* I don't think we need the full product of "MPI Implementations" times "Systems". We could just use OpenMPI by default and additionally test MPICH on one selected system (e.g. Ubuntu 18.04) ? 
* We also don't have "Features" times "MPI Implementations". 
So, OpenMPI: 24 + MPICH 4 ("Build Type", "PETSc on/off) = 28.

What do we do with PETSc? Against which version should we test on which system? I think it would be good to test 3.6 built from source with this 64bit option enabled (that should work, correct?) and also test 3.12 built from source (without this option) as well as using the system installation (Ubuntu 20.04).

**Comment by MakisH** [comment]
> What do we do with PETSc? Against which version should we test on which system? I think it would be good to test 3.6 built from source with this 64bit option enabled (that should work, correct?) and also test 3.12 built from source (without this option) as well as using the system installation (Ubuntu 20.04).

I think that for the usual suspects (i.e. PETSc) we should test the minimum required (past), the recommended (present), and the latest (future) one. We could then have:

- [past] Ubuntu 18.04 with PETSc 3.6 from source
- [present] Ubuntu 20.04 with PETSc 3.12 from APT
- [future] Arch with latest PETSc (updating e.g. once per day/week)

As "past" we can also use CentOS 7.

For "future", we could also explicitly have preCICE built with Spack. This would actually be a more common use case than Arch. Both are fetching the latest versions by default, it is only a matter of what gets updated more often. In Spack we can also always get the latest master of each dependency.

**Comment by fsimonis** [comment]
> As "future", we could also explicitly have preCICE built with Spack. This would actually be a more common use case than Arch.

True, the only downside is that it requires to build everything from source. Arch has at least prebuild packages.

**Comment by ajaust** [comment]
I would vote to keep at least some tests with Clang in the test suite. At least keep you code Clang-compatible. :)

New supercomputers with AMD Epyc architecture might come with [AOCC](https://developer.amd.com/amd-aocc/) (AMD Optimizing C/C++ Compiler) which is based on Clang. One example would be Hawk at the HLRS. The explicity mention the compiler in their [documentation](https://kb.hlrs.de/platforms/index.php/Compiler(Hawk)#AOCC). I cannot say though whether it is actually beneficial for performance to use the compiler. 

I am planning on using preCICE on Hawk soon when it becomes a bit more stable.

**Comment by fsimonis** [comment]
The result of the discussion is to test:

On `{Ubuntu 18.04 LTS, Arch Linux}`
  * `{Bare, MPI, MPI+PETSc} x {Debug, Release}` with GCC _6 jobs_
  * `{MPI+PETSc} x {Debug, Release}` with Clang _2 jobs_

This results in 16 jobs per run.

**Comment by fsimonis** [comment]
We could generate a complete build environment using MPICH and SPACK under CentOS 7 following [this guide](https://spack.readthedocs.io/en/latest/containers.html#environment-configuration).

**Comment by fsimonis** [comment]
We could add the Intel Compiler to the matrix by using the official [Intel One API - hpckit](https://hub.docker.com/r/intel/oneapi-hpckit) docker image.

**Comment by fsimonis** [comment]
I added Fedora 34 as it covers the workstation segment and has MPICH packages available.

This weakens the rational for separately testing spack, as we cover MPICH with Fedora and "bleeding edge" with archlinux.
I am also not sure if there is extra value in testing CentOS as both Fedora and CentOS are based on RHEL and the future of CentOS looks a bit grim.

**Comment by MakisH** [comment]
To my understanding, [Rocky Linux](https://rockylinux.org/) is the de facto replacement for CentOS in the server segment, i.e., the distribution that many shared computer systems will be based upon in the near future. We can test that one, instead. I have already checked and our instructions for CentOS 8 apply directly to Rocky Linux 8: https://precice.org/installation-source-dependencies.html

**Comment by MakisH** [comment]
@fsimonis I assume we can already close this by now. We are already using GitHub Actions and there is already a quite elaborate matrix: https://github.com/precice/precice/blob/develop/.github/workflows/build-and-test.yml#L37

I guess you kept it open to keep optimizing, but I think optimizations can come with separate issues/PRs.
